utilities/include/vec3.hpp

```diff
@@ -1,212 +1,59 @@
 #ifndef VEC3_HPP
 #define VEC3_HPP
 
-
 #include <iostream>
-#include <cmath> //constexpr is compiler specific
-class vec3{
-    float x,y,z;
-    public:
-        constexpr vec3() : x {0}, y {0}, z {0} {}
-
-        //Cartesian coordinate Constructor
-        constexpr vec3(float X,float Y,float Z) : x {X}, y {Y}, z {Z} {}
-
-        //norm function
-        #ifndef __GNUC__
-
-        float norm() const {
-            return sqrt(x * x + y * y + z * z);
-        }
-
-        #endif
-
-        #ifdef __GNUC__
-        
-        constexpr float norm() const {
-            return sqrt(x * x + y * y + z * z);
-        }
-
-        #endif
-
-
-
-        //unit vector function
-        constexpr vec3 unit() const{
-            return *this/this->norm();
-        }
-
-        //getters
-        constexpr float X() const{
-            return x;
-        }
+#include <cmath>
+#include <type_traits>
+
+#include "vec4.hpp"
+#include "config.hpp"
+
+class alignas(ALIGN_WIDTH) Vec3 : public Vec4{
+    private:
+                
+        /*
+        structs as_index,indices are used to generate appropriate xsimd::batch_constant 
+        to be used for xsimd::swizzle function used inside cross product. 
+        */
+        template <class T>
+        struct as_index
+        {
+            using type = xsimd::as_unsigned_integer_t<T>;
+        };
+        struct indices
+        {
+            static constexpr unsigned get(unsigned i, unsigned n)
+            {
+                if(i==n-1) return i;
+                else return (i+1)%(n-1);
+            }
+        };
 
-        constexpr float Y() const{
-            return y;
-        }
+    public:
+        constexpr Vec3(Utype x,Utype y,Utype z) : Vec4{x,y,z,0} {}
+        constexpr explicit Vec3(xsimd::batch<Utype,UArch> x) : Vec4{x} {}
         
-        constexpr float Z() const{
-            return z;
+        // cross-product
+        constexpr Vec3 cross(const Vec3& v2) const{
+            if (std::is_constant_evaluated()){
+                return Vec3((*this).y()*v2.z() - (*this).z()*v2.y(),(*this).z()*v2.x() - (*this).x()*v2.z(),(*this).x()*v2.y() - (*this).y()*v2.x());
```

Comment:
Put this on multiple lines, with alignment. If you have it as a `friend` function, then you would have these many instances of `*this`.

Reply:
But take this as a suggestion, not as something you must do. Consistency (having everything as `friend`s, or nothing as `friend`s), is most important.